library/request_queue.h

```diff
@@ -0,0 +1,96 @@
+#ifndef IOREMAP_ELLIPTICS_REQUEST_QUEUE_HPP
+#define IOREMAP_ELLIPTICS_REQUEST_QUEUE_HPP
+
+#include "elliptics.h"
+#include "murmurhash.h"
+
+#ifdef __cplusplus
+#include <unordered_map>
+#include <condition_variable>
+#include <mutex>
+#if __GNUC__ == 4 && __GNUC_MINOR__ < 5
+#  include <cstdatomic>
+#else
+#  include <atomic>
+#endif
+
+
+namespace std
+{
+	template<>
+	struct hash<dnet_id>
+	{
+		typedef dnet_id argument_type;
+		typedef std::size_t result_type;
+
+		result_type operator()(const argument_type &key) const
+		{
+			return MurmurHash64A(reinterpret_cast<const char *>(&key), sizeof(dnet_id), 0);
+		}
+	};
+}
+
+inline bool operator == (const dnet_id &lhs, const dnet_id &rhs)
+{
+	return !dnet_id_cmp(&lhs, &rhs);
+}
+
+struct dnet_locks_entry
+{
+	std::condition_variable unlock_event;
+};
+
+class dnet_request_queue
```

Comment:
You absolutely **have** to write a detailed description of the new locking design.
This is a substantial part of the processing engine, and I've already spotted issues with mutexes.